tests/sentry/search/events/test_builder.py

```diff
@@ -565,3 +568,208 @@ def test_disable_auto_aggregation(self):
+        with self.assertRaisesRegex(
+            InvalidSearchQuery,
+            re.escape("Tag value was not found"),
+        ):
+            MetricsQueryBuilder(
+                self.params,
+                "transaction:something_else",
+                selected_columns=["transaction", "project", "p95(transaction.duration)"],
+            )
+
+    def test_missing_transaction_index_in_filter(self):
+        with self.assertRaisesRegex(
+            InvalidSearchQuery,
+            re.escape("Tag value was not found"),
+        ):
+            MetricsQueryBuilder(
+                self.params,
+                "transaction:[something_else, something_else2]",
+                selected_columns=["transaction", "project", "p95(transaction.duration)"],
+            )
+
+    def test_project_filter(self):
+        query = MetricsQueryBuilder(
+            self.params,
+            f"project:{self.project.slug}",
+            selected_columns=["transaction", "project", "p95(transaction.duration)"],
+        )
+        self.assertCountEqual(
+            query.where,
+            [*self.default_conditions, Condition(Column("project_id"), Op.EQ, self.project.id)],
+        )
+
+    def test_granularity(self):
+        # Need to pick granularity based on the period
+        def get_granularity(start, end):
+            params = {
+                "organization_id": self.organization_id,
+                "project_id": self.projects,
+                "start": start,
+                "end": end,
+            }
+            query = MetricsQueryBuilder(params)
+            return query.granularity.granularity
+
+        # If we're doing atleast day and its midnight we should use the daily bucket
+        start = datetime.datetime(2015, 5, 18, 0, 0, 0, tzinfo=timezone.utc)
+        end = datetime.datetime(2015, 5, 18, 0, 0, 0, tzinfo=timezone.utc)
+        assert get_granularity(start, end) == 86400, "A day at midnight"
+
+        # If we're on the start of the hour we should use the hour granularity
+        start = datetime.datetime(2015, 5, 18, 23, 0, 0, tzinfo=timezone.utc)
+        end = datetime.datetime(2015, 5, 20, 1, 0, 0, tzinfo=timezone.utc)
+        assert get_granularity(start, end) == 3600, "On the hour"
+
+        # Even though this is >24h of data, because its a random hour in the middle of the day to the next we use minute
+        # granularity
+        start = datetime.datetime(2015, 5, 18, 10, 15, 1, tzinfo=timezone.utc)
+        end = datetime.datetime(2015, 5, 19, 15, 15, 1, tzinfo=timezone.utc)
+        assert get_granularity(start, end) == 60, "A few hours, but random minute"
+
+        # Less than a minute, no reason to work hard for such a small window, just use a minute
+        start = datetime.datetime(2015, 5, 18, 10, 15, 1, tzinfo=timezone.utc)
+        end = datetime.datetime(2015, 5, 19, 10, 15, 34, tzinfo=timezone.utc)
+        assert get_granularity(start, end) == 60, "less than a minute"
```

Comment:
@evanh this test is probably the clearest way to see how I'm planning to determine granularity.
I have a TODO to allow more flexibility on using a larger granularity (eg 15:59 -> 17:01 could likely still use the 3600 bucket) but I'm hoping this is a reasonable starting place.
I've intentionally omitted the 10s bucket since I don't see us having a valid usecase for that

Reply:
Looks good, thanks!